### ISSUE TITLE
Can't pickle the parallel function 

### ISSUE THREAD

**Issue by simonzcaiman** [joblib/joblib#543]
Hi guys, 

I have recently encountered an issue about not being able to pickle the function. 
For example , I want to run a function read_csv() and this read_csv is a class method.
The following function will be called in another class method in the same class. 
Parallel(n_jobs=20, verbose=50, timeout=100.0)(delayed(read_csv)(x) for x in pop)
And I run this script with Parallel for loop, it works perfect. 

Now I have this script written in a class format , such that I can import it from my other scripts.
However, if I have another script, import my module which contains the parallel for loop in it,and I created a class object and called the class method which will then call the parallel for loop, there will be pickle error saying cannot pickle the read_csv function. I'm wondering is it necessary to have the parallel for loop function in the main thread/script? 

Thanks guys

**Comment by lesteve** [comment]
Can you provide a stand-alone snippet to reproduce the problem ? Please read https://stackoverflow.com/help/mcve.

**Comment by jason-neal** [comment]
I have a similar issue with not pickling the function, although it doesn't involve classes/other modules.

The function I wish to parallelize has many extra args and kwargs that it needs apart from the parameter to parallelize over.  
So that I don't have to specify the all the extra parameters in the Parallel line I created a wrapper function that gets all the other parameters from the outer scope, with just the parallelized parameters as inputs.

The commented out joblib line does work, where as the one the uses the wrapper function does not and raises t
`AttributeError: Can't pickle local object 'parallel_work.<locals>.func_wrapper`

Here is the code to reproduce this.
```
from joblib import Parallel, delayed

def main():
    res = parallel_work()
    print(res)

def parallel_work():
    params = range(10)

    arg = "hello "
    kwarg = "world"

    def func_wrapper(num, param):
        """Take only parallel parameters."""
        return complex_function(num, param, arg, kwarg=kwarg)

    vals = Parallel(n_jobs=-2)(delayed(func_wrapper)(ii, param) for ii, param in enumerate(params))
    # vals = Parallel(n_jobs=-2)(delayed(complex_function)(ii, param, arg, kwarg="world")
    #                            for ii, param in enumerate(params))

    return vals

def complex_function(num, param, arg, kwarg=""):
    """Long complicated function to parallize.
    With many other arg and kwarg parameters.
    (Actually this saves to individual files with
    names derived from args and "num").
    """
    return "{0} = {1}{2}".format(num, arg * param, kwarg)


if __name__ == "__main__":
    main()

```

**Comment by ogrisel** [comment]
@jason-neal @simonzcaiman could you try again with joblib master? We now use a vendored copy of cloudpickle which should solve your issues. In particular I do not get the `AttributeError` exception on @jason-neal's snippet.

**Comment by jason-neal** [comment]
That seems to have solved my issue. Thanks!

**Comment by lesteve** [comment]
Thanks @jason-neal. Let's close this one then.

**Comment by maxibor** [comment]
That fixed the same issue for me too !

**Comment by CompPhysChris** [comment]
My group is waiting for this update as well.  We have many users who expand on existing analysis notebooks and it would be great if they could define their functions in the notebook for parallel execution.  Is there a timetable for the next release?

**Comment by jason-neal** [comment]
You can install the master branch with the fix using 

```
git clone git://github.com/joblib/joblib.git
cd joblib
python setup.py install
```

It is not as nice to use as `pip install joblib` but it works.

**Comment by lesteve** [comment]
FWIW you can install directly from master like this:
```
pip install git+https://github.com/joblib/joblib
```

You can also explicitly specify a given commit if you'd rather not have your joblib code change when we update master. You can do a similar thing in requirements.txt I am pretty sure.

> Is there a timetable for the next release?

Short answer: not really. Longer answer: there are quite a few major changes that went into master since last release and it is quite likely that it will take some time to settle down. If you do use master and notice problems while using it, please do report them, it is invaluable feed-back to squash regressions before the next release!

**Comment by CompPhysChris** [comment]
Thanks for the info.

**Comment by keandersen** [comment]
Is there a commit associated with this fix?
Is it included in the 0.12 release?

**Comment by bilalghanem** [comment]
I had the same case of @jason-neal which explained above. My code is working in one PC while not in another. I checked the version of the library and I found that I got this error when I run the code on the PC which has the version 0.14 while it works with the 0.12.1 version. Thus, I upgraded the version and it works now.
